[1.20.6] Add BlockTagIngredient for when a item tag cannot be used/doesn't exist

IThundxr opened the pull request:
In some certain use cases using the normal tag ingredient with a ItemTag isn't always possible as a tag for that can not exist. A good example of this is `minecraft:convertable_to_mud` this tag only exists as a block tag with no ItemTag that can be used instead.

Example usage: <https://github.com/Creators-of-Create/Create/blob/mc1.20.1/dev/src/generated/resources/data/create/recipes/mixing/mud_by_mixing.json>

The type id for this is "neoforge:block_tag"

IThundxr committed:
Implement BlockTagIngredient

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=976) <br/>All committers have signed the CLA.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:
```diff
@@ -0,0 +1,93 @@
+        this.tag = tag;
+    }
+
+    protected void dissolve() {
```
This shouldn't be necessary, caching this is already handled by Ingredient. But is that cache accessible from the custom ingredient? Maybe not...

IThundxr replied:
I don't think i can access that cache from here

KnightMiner replied:
Given this list is used in `test`, it might still be an issue as this cache is never invalidated. I suppose its only a problem if someone statically stores a block tag ingredient though.

Technici4n commented on src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:
```diff
@@ -0,0 +1,93 @@
+import net.minecraft.world.level.block.Block;
+import net.neoforged.neoforge.common.NeoForgeMod;
+
+/**
```
Javadoc needs a bit of formatting work (@link, @code, etc).

Technici4n commented on src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:
```diff
@@ -0,0 +1,93 @@
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
```
Prefer the JetBrains one.

IThundxr committed:
chore: fix reviews

- Format javadoc a bit better
- Switch to jetbrains @Nullable annotation
- Add nullcheck to make IDE's not angry

IThundxr committed:
Fix Spotless being angry

embeddedt commented on src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:
```diff
@@ -0,0 +1,94 @@
+    @Override
+    public Stream<ItemStack> getItems() {
+        dissolve();
+        return itemStacks != null ? Stream.of(itemStacks) : Stream.empty();
```
`itemStacks` will never be `null` after a call to `dissolve`

IThundxr replied:
That's kinda of just to make intellij not complain, could replace it with a noinspection if you'd prefer that instead

IThundxr committed:
ok intellij, go home

IThundxr commented:
todo: implement hashCode & equals

KnightMiner commented on src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:
```diff
@@ -0,0 +1,94 @@
+            return stack.isEmpty();
+        }
+
+        for (ItemStack itemStack : itemStacks) {
+            if (itemStack != null && itemStack.is(stack.getItem())) {
+                return true;
+            }
+        }
```
Why not just store a lazily initialized set of values to check against? Would be notably faster. Can use the set to construct `getItems()`

KnightMiner commented on src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:
```diff
@@ -0,0 +1,94 @@
+        }
+
+        dissolve();
+        if (itemStacks.length == 0) {
+            return stack.isEmpty();
+        }
```
I know vanilla ingredients do this, but I don't think modded ingredients want this. Neo notably patches the vanilla ingredient to sync a dummy item to the client for empty ingredients to ensure an empty tag (e.g. unused mod compat) does not become "fill with nothing".

IThundxr committed:
important branding

IThundxr committed:
fix empty tags returning true in test

IThundxr committed:
fix @NotNull

Technici4n commented on .idea/icon.svg:
Shouldn't this be left for another commit? :P

Technici4n reviewed (dismissed):
Besides that `.idea/icon.svg` I'm happy. Maybe that can be sent in a separate PR alongside `.editorconfig`?

IThundxr commented:
i don't really have much interest in maintaining a editorconfig pr, i'll make a separate one for the intellij icon

IThundxr committed:
Delete .idea directory

Technici4n commented on tests/src/main/java/net/neoforged/neoforge/debug/crafting/IngredientTests.java:
```diff
@@ -66,6 +68,37 @@
+            @Override
+            protected void buildRecipes(RecipeOutput output) {
+                ShapelessRecipeBuilder.shapeless(RecipeCategory.MISC, Items.MUD)
+                        .requires(new BlockTagIngredient(BlockTags.CONVERTABLE_TO_MUD).toVanilla())
```
Sorry I missed this in my previous review: Can you wrap the ingredient in a `new TestEnabledIngredient(`? (See example below).

Technici4n commented on src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:
```diff
@@ -0,0 +1,110 @@
+
+        dissolve();
+        for (ItemStack itemStack : itemStacks) {
+            if (itemStack != null && itemStack.is(stack.getItem())) {
```
This null check seems unnecessary?

IThundxr committed:
address reviews

IThundxr committed:
Merge remote-tracking branch 'origin/feat/blocktag-ingredient' into feat/blocktag-ingredient

IThundxr committed:
apply spotless

pupnewfster commented on src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java:
```diff
@@ -0,0 +1,110 @@
+            }
+
+            if (list.isEmpty()) {
+                ItemStack itemStack = new ItemStack(net.minecraft.world.level.block.Blocks.BARRIER);
```
This doesn't need to be a FQN because we are not in a patch.

IThundxr committed:
address reviews
